lib/EarlyFunctionAnalysis/DetectABI.cpp

```diff
@@ -488,6 +494,99 @@ void DetectABI::interproceduralPropagation() {
   }
 }
 
+void DetectABI::propagatePrototypes() {
+  for (model::Function &Function : Binary->Functions()) {
+    const auto &Entry = Function.Entry();
+
+    auto &Summary = Oracle.getLocalFunction(Entry);
+    auto &ABI = Summary.ABIResults;
+    auto &CFG = Summary.CFG;
+    auto &WrittenRegisters = Summary.WrittenRegisters;
+
+    const bool HasNoArguments = ABI.ArgumentsRegisters.size() == 0;
+    const bool HasNoReturns = ABI.FinalReturnValuesRegisters.size() == 0;
+    const bool IsSingleNode = CFG.size() == 1;
+
+    if (HasNoReturns and HasNoArguments and IsSingleNode) {
+      auto &Block = *CFG.begin();
+
+      if (Block.Successors().size() == 1) {
+        auto &Successor = *Block.Successors().begin();
+
+        // Select new prototype for wrapper function
+        if (auto Call = dyn_cast<efa::CallEdge>(Successor.get())) {
+          auto Prototype = getPrototype(*Binary, Entry, Block, *Call);
+
+          using abi::FunctionType::Layout;
+          // Get layout of wrapped function
+          Layout CalleeLayout = Layout::make(Prototype);
+
+          // Verify that wrapper function:
+          //  - don't write stack pointer
+          //  - don't write to callee arguments
+          //  - every store instruction writes to registers (not memory)
+          auto *BB = GCBI.getBlockAt(Block.ID().start());
+
+          llvm::GlobalVariable *StackPointer = nullptr;
+          {
+            const auto &Architecture = Binary->Architecture();
+            auto SP = model::Architecture::getStackPointer(Architecture);
+            llvm::StringRef StackPointerName = model::Register::getCSVName(SP);
+            StackPointer = M.getGlobalVariable(StackPointerName);
+          }
+
+          // Check if wrapper writes to Stack Pointer
+          const bool WritesSP = WrittenRegisters.count(StackPointer) > 0;
+
+          using std::ranges::count_if;
+          auto IsWritingToCalleeArg = [this,
```

Comment:
* `IsWritingToCalleeArg` -> `IsWrittenByCaller` or...
* Rewrite creating a set of `model::Register` and see if intersection is empty